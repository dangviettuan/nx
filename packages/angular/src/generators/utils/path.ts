import { joinPathFragments } from '@nrwl/devkit';
import { basename, dirname, normalize, relative } from 'path';

export function pathStartsWith(path1: string, path2: string) {
  path1 = normalize(path1).replace(/\\/g, '/');
  path2 = normalize(path2).replace(/\\/g, '/');

  return path1.startsWith(path2);
}

export function getRelativeImportToFile(
  sourceFilePath: string,
  targetFilePath: string
): string {
  const relativeDirToTarget = relative(
    dirname(sourceFilePath),
    dirname(targetFilePath)
  );

  return `./${joinPathFragments(
    relativeDirToTarget,
    basename(targetFilePath, '.ts')
  )}`;
}
