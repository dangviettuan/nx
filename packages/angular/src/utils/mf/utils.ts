import {
  applyAdditionalShared,
  applySharedFunction,
  createProjectGraphAsync,
  getDependentPackagesForProject,
  mapRemotes,
  mapRemotesForSSR,
  ModuleFederationConfig,
  ProjectGraph,
  readCachedProjectGraph,
  SharedLibraryConfig,
  sharePackages,
  shareWorkspaceLibraries,
} from '@nrwl/devkit';

export function applyDefaultEagerPackages(
  sharedConfig: Record<string, SharedLibraryConfig>
) {
  const DEFAULT_PACKAGES_TO_LOAD_EAGERLY = [
    '@angular/localize',
    '@angular/localize/init',
  ];
  for (const pkg of DEFAULT_PACKAGES_TO_LOAD_EAGERLY) {
    if (!sharedConfig[pkg]) {
      continue;
    }

    sharedConfig[pkg] = { ...sharedConfig[pkg], eager: true };
  }
}

export const DEFAULT_NPM_PACKAGES_TO_AVOID = ['zone.js', '@nrwl/angular/mf'];
export const DEFAULT_ANGULAR_PACKAGES_TO_SHARE = [
  '@angular/animations',
  '@angular/common',
];

export async function getModuleFederationConfig(
  mfConfig: ModuleFederationConfig,
  determineRemoteUrl: (remote: string) => string,
  options: { isServer: boolean } = { isServer: false }
) {
  let projectGraph: ProjectGraph;
  try {
    projectGraph = readCachedProjectGraph();
  } catch (e) {
    projectGraph = await createProjectGraphAsync();
  }

  if (!projectGraph.nodes[mfConfig.name]?.data) {
    throw Error(
      `Cannot find project "${mfConfig.name}". Check that the name is correct in module-federation.config.js`
    );
  }

  const dependencies = getDependentPackagesForProject(
    projectGraph,
    mfConfig.name
  );

  if (mfConfig.shared) {
    dependencies.workspaceLibraries = dependencies.workspaceLibraries.filter(
      (lib) => mfConfig.shared(lib.name, {})
    );
    dependencies.npmPackages = dependencies.npmPackages.filter((pkg) =>
      mfConfig.shared(pkg, {})
    );
  }

  const sharedLibraries = shareWorkspaceLibraries(
    dependencies.workspaceLibraries
  );

  const npmPackages = sharePackages(
    Array.from(
      new Set([
        ...DEFAULT_ANGULAR_PACKAGES_TO_SHARE,
        ...dependencies.npmPackages.filter(
          (pkg) => !DEFAULT_NPM_PACKAGES_TO_AVOID.includes(pkg)
        ),
      ])
    )
  );

  DEFAULT_NPM_PACKAGES_TO_AVOID.forEach((pkgName) => {
    if (pkgName in npmPackages) {
      delete npmPackages[pkgName];
    }
  });

  const sharedDependencies = {
    ...sharedLibraries.getLibraries(),
    ...npmPackages,
  };

  applyDefaultEagerPackages(sharedDependencies);
  applySharedFunction(sharedDependencies, mfConfig.shared);
  applyAdditionalShared(
    sharedDependencies,
    mfConfig.additionalShared,
    projectGraph
  );

  const mapRemotesFunction = options.isServer ? mapRemotesForSSR : mapRemotes;
  const mappedRemotes =
    !mfConfig.remotes || mfConfig.remotes.length === 0
      ? {}
      : mapRemotesFunction(mfConfig.remotes, 'mjs', determineRemoteUrl);
  return { sharedLibraries, sharedDependencies, mappedRemotes };
}
