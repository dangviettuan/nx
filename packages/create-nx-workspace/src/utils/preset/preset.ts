export enum Preset {
  Apps = 'apps',
  Empty = 'empty', // same as apps, deprecated
  Core = 'core', // same as npm, deprecated
  NPM = 'npm',
  TS = 'ts',
  WebComponents = 'web-components',
  AngularMonorepo = 'angular-monorepo',
  AngularStandalone = 'angular-standalone',
  ReactMonorepo = 'react-monorepo',
  ReactStandalone = 'react-standalone',
  ReactNative = 'react-native',
  Expo = 'expo',
  NextJs = 'next',
  Nest = 'nest',
  Express = 'express',
  React = 'react',
  Angular = 'angular',
  NodeStandalone = 'node-standalone',
}

/**
 * This function is used to check if a preset is a known Nx preset.
 * @param preset
 * @returns true if the preset is a known Nx preset, false otherwise.
 */
export function isKnownPreset(preset: string): preset is Preset {
  return Object.values(Preset).includes(preset as Preset);
}
