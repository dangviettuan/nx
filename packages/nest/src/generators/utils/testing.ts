import { addProjectConfiguration, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';

export function createTreeWithNestApplication(appName: string): Tree {
  const tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  addProjectConfiguration(tree, appName, {
    root: `apps/${appName}`,
    sourceRoot: `apps/${appName}/src`,
    projectType: 'application',
    targets: {},
  });
  return tree;
}
