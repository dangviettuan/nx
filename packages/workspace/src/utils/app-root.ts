// TODO(v15): Remove this file
export { appRootPath, workspaceRoot } from '@nrwl/devkit';
