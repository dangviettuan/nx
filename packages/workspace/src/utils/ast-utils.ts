import { getSourceNodes } from '../utilities/typescript/get-source-nodes';

export { findNodes } from '../utilities/typescript/find-nodes'; // TODO(v16): Remove this
export { getSourceNodes } from '../utilities/typescript/get-source-nodes';
