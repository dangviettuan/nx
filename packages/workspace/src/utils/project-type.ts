export enum ProjectType {
  Application = 'application',
  Library = 'library',
}

export function projectRootDir(projectType: ProjectType) {
  if (projectType == ProjectType.Application) {
    return 'apps';
  } else if (projectType == ProjectType.Library) {
    return 'libs';
  }
}

export function projectDir(projectType: ProjectType) {
  if (projectType == ProjectType.Application) {
    // apps/test-app/src/app
    return 'app';
  } else if (projectType == ProjectType.Library) {
    // libs/test-lib/src/lib
    return 'lib';
  }
}
