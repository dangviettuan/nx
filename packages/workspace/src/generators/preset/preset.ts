import {
  installPackagesTask,
  names,
  readNxJson,
  Tree,
  updateNxJson,
} from '@nrwl/devkit';
import { Schema } from './schema';
import { Preset } from '../utils/presets';

export async function presetGenerator(tree: Tree, options: Schema) {
  options = normalizeOptions(options);
  const presetTask = await createPreset(tree, options);
  return async () => {
    installPackagesTask(tree);
    if (presetTask) await presetTask();
  };
}

export default presetGenerator;

async function createPreset(tree: Tree, options: Schema) {
  if (options.preset === Preset.Empty || options.preset === Preset.Apps) {
    return;
  } else if (options.preset === Preset.AngularMonorepo) {
    const {
      applicationGenerator: angularApplicationGenerator,
    } = require('@nrwl' + '/angular/generators');

    return angularApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      standalone: options.standaloneApi,
      routing: options.routing,
    });
  } else if (options.preset === Preset.AngularStandalone) {
    const {
      applicationGenerator: angularApplicationGenerator,
    } = require('@nrwl' + '/angular/generators');

    return angularApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      routing: options.routing,
      rootProject: true,
      standalone: options.standaloneApi,
    });
  } else if (options.preset === Preset.ReactMonorepo) {
    const {
      applicationGenerator: reactApplicationGenerator,
    } = require('@nrwl' + '/react');

    return reactApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      bundler: options.bundler ?? 'webpack',
    });
  } else if (options.preset === Preset.ReactStandalone) {
    const {
      applicationGenerator: reactApplicationGenerator,
    } = require('@nrwl' + '/react');

    return reactApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      rootProject: true,
      bundler: options.bundler ?? 'vite',
      e2eTestRunner: 'cypress',
      unitTestRunner: options.bundler === 'vite' ? 'vitest' : 'jest',
    });
  } else if (options.preset === Preset.NextJs) {
    const { applicationGenerator: nextApplicationGenerator } = require('@nrwl' +
      '/next');

    return nextApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
    });
  } else if (options.preset === Preset.WebComponents) {
    const { applicationGenerator: webApplicationGenerator } = require('@nrwl' +
      '/web');

    return webApplicationGenerator(tree, {
      name: options.name,
      style: options.style,
      linter: options.linter,
      bundler: 'vite',
    });
  } else if (options.preset === Preset.Nest) {
    const { applicationGenerator: nestApplicationGenerator } = require('@nrwl' +
      '/nest');

    return nestApplicationGenerator(tree, {
      name: options.name,
      linter: options.linter,
    });
  } else if (options.preset === Preset.Express) {
    const {
      applicationGenerator: expressApplicationGenerator,
    } = require('@nrwl' + '/express');
    return expressApplicationGenerator(tree, {
      name: options.name,
      linter: options.linter,
    });
  } else if (options.preset === Preset.ReactNative) {
    const { reactNativeApplicationGenerator } = require('@nrwl' +
      '/react-native');
    return reactNativeApplicationGenerator(tree, {
      name: options.name,
      linter: options.linter,
      e2eTestRunner: 'detox',
    });
  } else if (options.preset === Preset.Expo) {
    const { expoApplicationGenerator } = require('@nrwl' + '/expo');
    return expoApplicationGenerator(tree, {
      name: options.name,
      linter: options.linter,
      e2eTestRunner: 'detox',
    });
  } else if (options.preset === Preset.TS) {
    const c = readNxJson(tree);
    const { initGenerator } = require('@nrwl' + '/js');
    c.workspaceLayout = {
      appsDir: 'packages',
      libsDir: 'packages',
    };
    updateNxJson(tree, c);
    return initGenerator(tree, {});
  } else if (options.preset === Preset.NodeStandalone) {
    const { applicationGenerator: nodeApplicationGenerator } = require('@nrwl' +
      '/node');
    return nodeApplicationGenerator(tree, {
      name: options.name,
      linter: options.linter,
      standaloneConfig: options.standaloneConfig,
      framework: options.framework,
      docker: options.docker,
      rootProject: true,
    });
  } else {
    throw new Error(`Invalid preset ${options.preset}`);
  }
}

function normalizeOptions(options: Schema): Schema {
  options.name = names(options.name).fileName;
  return options;
}
