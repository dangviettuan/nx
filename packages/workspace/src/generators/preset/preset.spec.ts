import { readProjectConfiguration, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { overrideCollectionResolutionForTesting } from '@nrwl/devkit/ngcli-adapter';
import { presetGenerator } from './preset';
import * as path from 'path';
import { Preset } from '../utils/presets';

describe('preset', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
    overrideCollectionResolutionForTesting({
      '@nrwl/workspace': path.join(
        __dirname,
        '../../../../workspace/generators.json'
      ),
      '@nrwl/angular': path.join(
        __dirname,
        '../../../../angular/generators.json'
      ),
      '@nrwl/linter': path.join(
        __dirname,
        '../../../../linter/generators.json'
      ),
      '@nrwl/nest': path.join(__dirname, '../../../../nest/generators.json'),
      '@nrwl/node': path.join(__dirname, '../../../../node/generators.json'),
      '@nrwl/jest': path.join(__dirname, '../../../../jest/generators.json'),
      '@nrwl/cypress': path.join(
        __dirname,
        '../../../../cypress/generators.json'
      ),
      '@nrwl/express': path.join(
        __dirname,
        '../../../../express/generators.json'
      ),
    });
  });

  afterEach(() => {
    overrideCollectionResolutionForTesting(null);
  });

  it(`should create files (preset = ${Preset.AngularMonorepo})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.AngularMonorepo,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.children('apps/proj')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/app')).toMatchSnapshot();
  }, 20000);

  it(`should create files (preset = ${Preset.WebComponents})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.WebComponents,
    });
    expect(tree.exists('/apps/proj/src/main.ts')).toBe(true);
  });

  it(`should create files (preset = ${Preset.ReactMonorepo})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactMonorepo,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.exists('/apps/proj/src/main.tsx')).toBe(true);
    expect(readProjectConfiguration(tree, 'proj').targets.serve).toBeDefined();
  });

  it(`should create files (preset = ${Preset.NextJs})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.NextJs,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.exists('/apps/proj/pages/index.tsx')).toBe(true);
  });

  it(`should create files (preset = ${Preset.Express})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.Express,
      linter: 'eslint',
    });

    expect(tree.exists('apps/proj/src/main.ts')).toBe(true);
    expect(tree.exists('apps/proj/.eslintrc.json')).toBe(true);
  });

  it('should create files (preset = react-native)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactNative,
      linter: 'eslint',
    });

    expect(tree.exists('/apps/proj/src/app/App.tsx')).toBe(true);
  });

  it(`should create files (preset = ${Preset.ReactStandalone} bundler = webpack)`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactStandalone,
      style: 'css',
      linter: 'eslint',
      bundler: 'webpack',
    });
    expect(tree.exists('webpack.config.js')).toBe(true);
    expect(
      readProjectConfiguration(tree, 'proj').targets.serve
    ).toMatchSnapshot();
  });

  it(`should create files (preset = ${Preset.ReactStandalone} bundler = vite)`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactStandalone,
      style: 'css',
      linter: 'eslint',
      bundler: 'vite',
    });
    expect(tree.exists('vite.config.ts')).toBe(true);
    expect(
      readProjectConfiguration(tree, 'proj').targets.serve
    ).toMatchSnapshot();
  });
});
