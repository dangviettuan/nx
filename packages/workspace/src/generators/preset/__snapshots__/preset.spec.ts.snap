// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset should create files (preset = angular-monorepo) 1`] = `
Array [
  "project.json",
  "src",
  "tsconfig.app.json",
  "tsconfig.editor.json",
  "tsconfig.json",
  ".eslintrc.json",
  "jest.config.ts",
  "tsconfig.spec.json",
]
`;

exports[`preset should create files (preset = angular-monorepo) 2`] = `
Array [
  "assets",
  "favicon.ico",
  "index.html",
  "styles.css",
  "app",
  "main.ts",
  "test-setup.ts",
]
`;

exports[`preset should create files (preset = angular-monorepo) 3`] = `
Array [
  "app.component.css",
  "app.component.html",
  "app.component.spec.ts",
  "app.component.ts",
  "app.module.ts",
  "nx-welcome.component.ts",
]
`;

exports[`preset should create files (preset = react-standalone bundler = vite) 1`] = `
Object {
  "configurations": Object {
    "development": Object {
      "buildTarget": "proj:build:development",
      "hmr": true,
    },
    "production": Object {
      "buildTarget": "proj:build:production",
      "hmr": false,
    },
  },
  "defaultConfiguration": "development",
  "executor": "@nrwl/vite:dev-server",
  "options": Object {
    "buildTarget": "proj:build",
  },
}
`;

exports[`preset should create files (preset = react-standalone bundler = webpack) 1`] = `
Object {
  "configurations": Object {
    "development": Object {
      "buildTarget": "proj:build:development",
    },
    "production": Object {
      "buildTarget": "proj:build:production",
      "hmr": false,
    },
  },
  "defaultConfiguration": "development",
  "executor": "@nrwl/webpack:dev-server",
  "options": Object {
    "buildTarget": "proj:build",
    "hmr": true,
  },
}
`;
